sql/logic_test.go

```diff
@@ -122,37 +122,76 @@ func (v *logicValue) Scan(value interface{}) error {
 // functions). We should work towards fixing that.
 type logicTest struct {
 	*testing.T
+	file *os.File
+	srv  *server.TestServer
+	// map of built clients. Needs to be persisted so that we can
+	// re-use them and close them all on exit.
+	clients map[string]*sql.DB
+	// client currently in use.
+	db *sql.DB
+}
+
+func (t *logicTest) close() {
+	if t.file != nil {
+		t.file.Close()
+		t.file = nil
+	}
+	if t.srv != nil {
+		t.srv.Stop()
+		t.srv = nil
+	}
+	if t.clients != nil {
+		for _, c := range t.clients {
+			c.Close()
+		}
+		t.clients = nil
+	}
+	t.db = nil
+}
+
+// setUser sets the DB client to the specified user, creating it if needed.
+func (t *logicTest) setUser(user string) {
+	if t.clients == nil {
+		t.clients = map[string]*sql.DB{}
+	}
+	if c, ok := t.clients[user]; ok {
+		t.db = c
+		return
+	}
+	db, err := sql.Open("cockroach", "https://"+user+"@"+t.srv.ServingAddr()+"?certs=test_certs")
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.clients[user] = db
+	t.db = db
 }
 
 func (t logicTest) run(path string) {
-	file, err := os.Open(path)
+	defer t.close()
+
+	var err error
+	t.file, err = os.Open(path)
```

Comment:
I think `file` can remain a local. It isn't necessary to place it in `logicTest`.